src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -3103,7 +3104,20 @@ private void BindArrayInitializerExpressions(InitializerExpressionSyntax initial
                 }
             }
 
-            return new BoundStackAllocArrayCreation(node, elementType, count, type: null, hasErrors: hasErrors || typeHasErrors);
+            TypeSymbol type = null;
+            if (!node.IsVariableDeclarationInitialization())
+            {
+                CheckFeatureAvailability(node, MessageID.IDS_FeatureRefStructs, diagnostics);
+                GetWellKnownTypeMember(Compilation, WellKnownMember.System_Span_T__ctor, diagnostics, syntax: node);
+
+                var spanType = GetWellKnownType(WellKnownType.System_Span_T, diagnostics, node);
+                if (!spanType.IsErrorType())
+                {
+                    type = spanType.Construct(elementType);
+                }
+            }
+
+            return new BoundStackAllocArrayCreation(node, elementType, count, type, hasErrors: hasErrors || typeHasErrors);
```

Comment:
Are there semantic model tests showing this new natural type (and in some error cases `null` type)?

Reply:
Semantic model is shortcircuited to report type and converted type always the same. That did not change. No need for new tests unless we change how semantic model reports these things.

---
In reply to: [139546808](https://github.com/dotnet/roslyn/pull/22162#discussion_r139546808) [](ancestors = 139546808)